### PR TITLE
re-implement chat in react

### PR DESCRIPTION
**Pull request by Connoropolous**
Devin says: let's make this PR only about porting the MapChat to react, with complete feature parity. Then we can do the chat upgrades later (so maybe we can get this done by new year's, and I can do the emoji next).
TODO:

- [ ] port ChatView to react
- [ ] fix all logic so ChatView.js is just a wrapper for the react components
- [ ] test the chat view to make sure it works the same as it used to
- [ ] add chat tests to metamaps-qa-steps.md

For the next PR, here's the chat upgrades I want to do:

- [ ] add missive/emoji-mart react emoji picker
- [ ] consider if it actually is OK to store emoji directly in the db
- [ ] if not, store emoji as shortcodes in the db
- [ ] make sure shortcodes and smileys are re-rendered as emoji if typed manually
- [ ] add logic to check for @ mentions in chat messages
- [ ] send @ mentions as an event on the realtime server
- [ ] potentially update the clients when there's an @ mention.

### CODE REVIEW

File: frontend/src/components/MapChat/index.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+          PARTICIPANTS
+          <div onClick={this.toggleVideosShowing} className={`video-toggle ${videosShowing ? '' : 'active'}`} />
+          <div onClick={this.toggleCursorsShowing} className={`cursor-toggle ${cursorsShowing ? '' : 'active'}`} />
+        </div>
+        <div className="participants">
+          <div className="conversation-live">
+            LIVE
+						<span className="call-action leave" onClick={this.props.leaveCall}>
+							LEAVE
+						</span>
```
Comment (Connoropolous) [inline frontend/src/components/MapChat/index.js]: these two divs are mutually exclusive. if you're in the call, you the first one, if youre not in the call you see the second one 'join'

**Comment by Connoropolous** [comment]
Agreed

File: frontend/src/components/MapChat/Participant.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  render() {
+    const { id, self, image, username, selfName, color } = this.props
+    return (
+			<div className={`participant participant-${id} ${self ? 'is-self' : ''}`}>
+				<div className="chat-participant-image">
+					<img src={image} style={{ border: `2px solid ${color}`}} />
+				</div>
+				<div className="chat-participant-name">
+					{username} {self ? '(me)' : ''}
+				</div>
```
Comment (Connoropolous) [inline frontend/src/components/MapChat/Participant.js]: Gotta add the 'pending' state in here again

File: frontend/src/Metamaps/Realtime/index.js [hunk 3]
```diff
@@ -237,14 +230,14 @@ let Realtime = {
     $('.collabCompass').remove()
     if (self.room) {
       self.room.leave()
-      self.room.chat.hide()
-      self.room.chat.close()
+      ChatView.hide()
+      ChatView.close()
```
Comment (devvmh) [inline frontend/src/Metamaps/Realtime/index.js]: awesome

File: frontend/src/Metamaps/Views/ChatView.js [hunk 4]
```diff
@@ -133,226 +32,181 @@ var Private = {
...
-  }
-}
-
-var Handlers = {
-  buttonClick: function() {
-    if (this.isOpen) this.close()
-    else if (!this.isOpen) this.open()
+  setNewMap: function(room) {
+    const self = ChatView
+    self.room = room
```
Comment (Connoropolous) [inline frontend/src/Metamaps/Views/ChatView.js]: Remove `room` it's not even used

**Comment by Connoropolous** [comment]
To test the Junto call related parts that need testing, I sign into two separate user accounts in normal and incognito chrome windows, and go to the same map. You have to run the real-time server. There are some things which can only be tested with three people because it behaves differently for the third person when theres already a conversation in progress. We can test those interactions on heroku.

**Pull request merged**